File: android/app/src/test/java/com/openrung/client/net/ExitNodeDirectoryTest.kt
package com.openrung.client.net

import com.openrung.client.model.RelayDescriptor
import com.openrung.client.model.RelayListResponse
import com.openrung.client.sampleRelay
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ExitNodeDirectoryTest {
    private fun geo(
        countryCode: String,
        country: String = countryCode,
        latitude: Double = 0.0,
        longitude: Double = 0.0,
    ) = ClientGeoInfo(
        ip = "203.0.113.1",
        country = country,
        countryCode = countryCode,
        city = "",
        asn = "",
        isp = "",
        organization = "",
        latitude = latitude,
        longitude = longitude,
    )

    private fun response(relays: List<RelayDescriptor>) =
        RelayListResponse(count = relays.size, serverTime = "2026-01-01T00:00:00Z", relays = relays)

    @Test
    fun groupsRelaysByCountryAndPlacesAtCentroid() = runTest {
        val relays = listOf(
            sampleRelay(id = "a", publicHost = "1.1.1.1"),
            sampleRelay(id = "b", publicHost = "2.2.2.2"),
            sampleRelay(id = "c", publicHost = "3.3.3.3"),
        )
        val geoByHost = mapOf(
            "1.1.1.1" to geo("JP"),
            "2.2.2.2" to geo("JP"),
            "3.3.3.3" to geo("US"),
        )

        val regions = ExitNodeDirectory(
            fetchRelays = { response(relays) },
            lookupGeo = { host -> geoByHost[host] },
        ).load()

        assertEquals(2, regions.size)
        // Sorted by node count descending → Japan (2 nodes) first.
        val japan = regions.first()
        assertEquals("JP", japan.countryCode)
        assertEquals("Japan", japan.countryName)
        assertEquals(2, japan.nodeCount)
        assertEquals(36.20, japan.latitude, 0.001)
        assertEquals(138.25, japan.longitude, 0.001)
    }

    @Test
    fun fallsBackToGeoCoordinatesForUnknownCountry() = runTest {
        val regions = ExitNodeDirectory(
            fetchRelays = { response(listOf(sampleRelay(publicHost = "9.9.9.9"))) },
            lookupGeo = { geo("ZZ", country = "Nowhere", latitude = 12.5, longitude = 34.0) },
        ).load()

        assertEquals(1, regions.size)
        assertEquals("Nowhere", regions.first().countryName)
        assertEquals(12.5, regions.first().latitude, 0.001)
        assertEquals(34.0, regions.first().longitude, 0.001)
    }

    @Test
    fun skipsRelaysWithMissingOrBlankGeo() = runTest {
        val regions = ExitNodeDirectory(
            fetchRelays = {
                response(
                    listOf(
                        sampleRelay(id = "a", publicHost = "5.5.5.5"),
                        sampleRelay(id = "b", publicHost = "6.6.6.6"),
                    ),
                )
            },
            lookupGeo = { host -> if (host == "5.5.5.5") geo("", country = "") else null },
        ).load()

        assertTrue(regions.isEmpty())
    }
}

File: desktop/frontend/src/core/net/exitNodeDirectory.test.ts
// Ported from openrung-mobile-app/__tests__/core/exitNodeDirectory.test.ts
// (Jest → vitest). Verifies grouping-by-country+city, node counts, ordering,
// and that ungeolocated relays stay off the map.
import { describe, expect, it } from 'vitest';
import { loadExitNodeDirectory } from './exitNodeDirectory';
import type { RelayDescriptor, RelayListResponse } from '../model/relay';

function relay(overrides: Partial<RelayDescriptor>): RelayDescriptor {
  return {
    id: Math.random().toString(36).slice(2),
    public_host: '1.2.3.4',
    public_port: 443,
    protocol: 'vless-reality-vision',
    client_id: 'uuid',
    reality_public_key: 'pk',
    short_id: 'sid',
    server_name: 'sni',
    flow: 'xtls-rprx-vision',
    exit_mode: 'direct',
    max_sessions: 8,
    max_mbps: 100,
    volunteer_version: '1',
    registered_at: '2026-07-05T00:00:00Z',
    last_heartbeat_at: '2026-07-06T00:00:00Z',
    expires_at: '2026-07-06T01:00:00Z',
    ...overrides,
  };
}

function response(relays: RelayDescriptor[]): RelayListResponse {
  return { count: relays.length, server_time: '2026-07-06T00:00:00Z', relays };
}

describe('loadExitNodeDirectory', () => {
  it('groups usable located relays by country+city and counts nodes', async () => {
    const regions = await loadExitNodeDirectory({
      fetchRelays: async () =>
        response([
          relay({ country_code: 'jp', country: 'Japan', city: 'Tokyo', latitude: 35.6, longitude: 139.7 }),
          relay({ country_code: 'JP', country: 'Japan', city: 'Tokyo', latitude: 35.6, longitude: 139.7 }),
          relay({ country_code: 'SG', country: 'Singapore', city: 'Singapore', latitude: 1.35, longitude: 103.8 }),
        ]),
    });

    expect(regions).toHaveLength(2);
    // Sorted by node count desc: Tokyo (2) before Singapore (1).
    expect(regions[0]).toMatchObject({ countryCode: 'JP', city: 'Tokyo', nodeCount: 2 });
    expect(regions[0].relays).toHaveLength(2);
    expect(regions[1]).toMatchObject({ countryCode: 'SG', nodeCount: 1 });
  });

  it('drops relays the broker has not geolocated', async () => {
    const regions = await loadExitNodeDirectory({
      fetchRelays: async () => response([relay({ country_code: undefined, latitude: undefined })]),
    });
    expect(regions).toHaveLength(0);
  });

  it('excludes unusable relays before grouping', async () => {
    const regions = await loadExitNodeDirectory({
      fetchRelays: async () =>
        response([
          relay({ protocol: 'other', country_code: 'JP', latitude: 35.6, longitude: 139.7 }),
        ]),
    });
    expect(regions).toHaveLength(0);
  });
});
